File: leetcode/70.Climbing Stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = {
            1: 1,
            2: 2,
            3: 3
        }

        def dp(n):
            if n == 1:
                return 1
            if (n == 2):
                return 2
            else:
                if n in memo:
                    return memo[n]
                memo[n] = dp(n-1) + dp(n-2)
                return memo[n]
        dp(n)
        return memo[n]


print(f"==>> Solution: {Solution().climbStairs(2)}")  # 2
# print(f"==>> Solution: {Solution().climbStairs(3)}")  # 3
print(f"==>> Solution: {Solution().climbStairs(4)}")  # 5

File: leetcode/70.Climbing Stairs.ts
const memo: { [key: number]: number } = {
	1: 1,
	2: 2,
	3: 3,
};
function climbStairs(n: number): number {
	return dp(n);
}

function dp(n: number): number {
	if (n === 1) return 1;
	if (n === 2) return 2;
	if (memo[n]) return memo[n];
	return (memo[n] = dp(n - 1) + dp(n - 2));
}

console.log('🚀 ~ climbStairs ~ climbStairs:', climbStairs(3));
